Remove expensive tests and interpolate out-of-the-box initial conditions

billsacks opened the pull request:
### Description of changes

1. Removed / reworked some expensive tests (see below for details)

2. Interpolated all out-of-the-box initial conditions, so that the
   out-of-the-box version is now compatible with our current
   configuration. The changes from before were (a) our standard
   configuration now uses the gx1v7 rather than gx1v6 land mask; (b)
   many inactive points are now absent in memory.

### Specific notes

Contributors other than yourself, if any: none

CTSM Issues Fixed (include github issue #):
- Resolves #312
- Partially addresses #275 (just a bit)

Are answers expected to change (and if so in what way)? YES
- Answers change for cases that still need to do interpolation from these
  out-of-the-box finidat files. See below for details.

Any User Interface Changes (namelist or namelist defaults changes)? No

Testing performed, if any:
- Ran some manual tests; see below for details

### Details on testing performed

- Compared `ncdump -h` of old vs. new initial conditions files

- Ran the following tests, with comparison against master; verified that
  these all pass and are bit-for-bit, that they use the new initial
  conditions files, and that they now have `use_init_interp` unset
  (whereas they used to have `use_init_interp` set to `.true.`); note
  that these cover 6 of the 7 new initial conditions files:

  ```
  SMS_Ld1.f09_g17.I1850Clm45BgcGs.cheyenne_intel.clm-default
  SMS_Ld1.f09_g17.I1850Clm45BgcCruGs.cheyenne_intel.clm-default
  SMS_Ld1.f09_g17.I1850Clm50Sp.cheyenne_intel.clm-default
  SMS_Ld1.f09_g17.I1850Clm50BgcCrop.cheyenne_intel.clm-ciso
  SMS_Ld1.f09_g17.I1850Clm50BgcCropCru.cheyenne_intel.clm-ciso
  SMS_Ld1.f09_g17.I1850Clm50SpCru.cheyenne_intel.clm-default
  ```

- To cover the last new initial conditions file, ran the following test
  (with the same checks as above: passes and bit-for-bit, checked
  finidat, and checked `use_init_interp`), with `./xmlchange
  LND_TUNING_MODE=clm5_0_cam6.0` (for the baseline for this test, I set
  `init_interp_method = 'use_finidat_areas'`, which I also used in
  interpolating this initial conditions file):
  
  ```
  SMS_Ld1.f09_g17.I1850Clm50BgcCrop.cheyenne_intel
  ```

### Details on answer changes

As noted above, these changes are bit-for-bit for configurations that
can now use the out-of-the-box initial conditions without
interpolation. However, answers change for configurations that still
require interpolation of the out-of-the-box initial conditions (e.g.,
because they are at a different resolution than the finidat file).

I expected these answer changes around the Caspian Sea, due to the fact
that we are now effectively interpolating twice from the original file:
once to produce the new out-of-the-box file (which includes an
interpolation from `gx1v6` to `gx1v7`, and so interpolates onto the
Caspian Sea region) and once to go from `f09_g17` to the target
resolution (rather than interpolating directly from `f09_g16` to the
target resolution).

However, when I ran a test with comparison to baselines -
`ERP_Ld5.f19_g17.I1850Clm50Bgc.cheyenne_intel.clm-default` - I saw
changes in regions further afield than simply around the Caspian
Sea. See the two attached images showing FSH diffs for this test
compared with master (one showing FSH at a reasonable data range
[`FSH_diffs_significant.png`], and one that changes the range to show
all points with differences [`FSH_diffs_all.png`]. It's possible that
the other differences are also due to the Caspian Sea change, but it's
also possible that there is some other cause here that I don't
understand.

<img width="1453" alt="fsh_diffs_significant" src="https://user-images.githubusercontent.com/6266741/51406739-34782e80-1b17-11e9-8430-376190c257d0.png">

<img width="1453" alt="fsh_diffs_all" src="https://user-images.githubusercontent.com/6266741/51406723-29250300-1b17-11e9-8bcd-71ef3214212d.png">

I feel it's probably not worth the time to dig into these differences
further, unless others feel it's important to do so. (It's possible
that, with further digging, we'd uncover some issue. But I suspect this
will take a few days of digging, and it's also quite possible that, in
the end, we'll discover that everything is working as intended.) I was
at least happy to see that differences are isolated to a relatively
small number of grid cells, giving me confidence that things haven't
completely changed with the interpolation.

### Details on changes to test list

I went through our most expensive tests in the test list. Some general
things I did were:

- Changing ERI tests to short SMS tests or coarser-resolution

- Shortening some tests

- Changing some tests to lower-resolution where it doesn't seem
  important to use high-resolution

- Removed some expensive tests for which there were very similar tests
  (or multiple tests that individually covered the various things
  covered in this one test) already in the test suite

I realized in the midst of doing this that much of the cost of (some of)
these tests was in initialization, not the run loop. Thus, shortening
these tests may not actually have much benefit, whereas using
already-interpolated out-of-the-box initial conditions may have a bigger
impact. (The timings that guided my work were from before I interpolated
the out-of-the-box initial conditions files.) However, it still seemed
worth shortening some of these unnecessarily-long high-resolution tests.

Specifically: these tests were removed / replaced:

- `ERI_D_Ld9.f09_g16.I1850Clm45BgcCruGs.cheyenne_intel.clm-default`:
  This is the only test of this compset (which is scientifically
  supported), and the only ERI Clm45 test. Changed
  `ERP_P36x2_D_Ld5.f10_f10_musgs.I1850Clm45Bgc.cheyenne_gnu.clm-default`
  (which is a duplicate of an intel test) to an `ERI_D_Ld9 test`, so we
  have an ERI Clm45 test. Then changed the above ERI test to
  `SMS_D_Ld1.f09_g17`

- `ERI_D_Ld9.f09_g16.I1850Clm50Sp.cheyenne_intel.clm-default`: want a
  test of this scientifically-supported compset at a
  scientifically-supported resolution: changed to `SMS_D_Ld1.f09_g17`

- `ERP_D_Ld9.f09_g16.I1850Clm50SpCru.cheyenne_intel.clm-default`: want a
  test of this scientifically-supported compset at a
  scientifically-supported resolution: changed to `SMS_D_Ld1.f09_g17`

- `ERI_Ld9.f09_g17.I1850Clm50Bgc.cheyenne_intel.clm-drydepnomegan`:
  changed this to have a coarse-resolution ERI test and a short,
  high-resolution SMS test: Changed
  `ERP_Ld5.f10_f10_musgs.I1850Clm50Bgc.cheyenne_gnu.clm-drydepnomegan`
  to `ERI_Ld9` (there's still an intel version of that ERP test),
  changed the above to `SMS_Ld1`.

- `ERI_Ld9.f09_g17.I2000Clm50BgcCruGs.cheyenne_intel.clm-cn_conly`:
  Deleted, but in order to keep a `cn_conly` test, added a
  coarse-resolution test of that testmod:
  `ERP_D_P36x2_Ld3.f10_g10_musgs.I2000Clm50BgcCruGs.cheyenne_intel.clm-cn_conly`

- `ERI_Ld9.f09_g17.I2000Clm50BgcCruGs.cheyenne_intel.clm-drydepnomegan`:
  deleted without replacement

- `ERP_D_Ld5.f19_g17_gl4.I1850Clm50BgcCropG.cheyenne_intel.clm-glcMEC_changeFlags`:
  deleted: redundant with
  `ERP_D_Ld5.f19_g17_gl4.I1850Clm50BgcCrop.cheyenne_intel.clm-glcMEC_changeFlags`

- `ERP_D_P720x1_Ld5.hcru_hcru.I2000Clm50BgcCruGs.cheyenne_intel`: In
  order to keep a test of this very high resolution, replaced with the
  very short
  `ERP_D_P720x1_Ld5.hcru_hcru.I2000Clm50BgcCruGs.cheyenne_intel`, which
  is cold start to avoid the cost of interpolating initial conditions.
  
- `ERP_Ld5.f09_g17.I1850Clm50BgcCropCru.cheyenne_intel.clm-ciso`: just
  shortened to `_Ld3`

- `ERS_D.f19_g17.I1850Clm50BgcCrop.cheyenne_intel.clm-reseedresetsnow`:
  changed to `f10_f10_musgs`
  
- `ERS_D_Ld3.f09_g17_gl4.I1850Clm50BgcCrop.cheyenne_intel.clm-clm50KitchenSink`,
  `SMS_Lm1.f09_g17_gl4.I1850Clm50Bgc.cheyenne_intel.clm-clm50KitchenSink`
  and
  `ERS_Ly3.f10_f10_musgs.I1850Clm50BgcCrop.cheyenne_intel.clm-clm50KSinkMOut`:
  we don't need clm50KitchenSink or clm50KSinkMOut any more, since they
  just have the clm50 options. However, one other thing this testmod did
  was to test human stress indices. To cover this, I have introduced a
  new `extra_outputs` testmod, and added
  `ERP_P36x2_D_Ld3.f10_f10_musgs.I1850Clm50BgcCrop.cheyenne_gnu.clm-extra_outputs`
  For now this testmod just turns on the extra human stress indices, but
  I could imagine adding more to this testmod later.
  
- `SMS_D.f09_g16.I1850Clm50BgcSpinup.cheyenne_intel.clm-cplhist`:
  shortened to `SMS_D_Ld1`.
  
- `SMS_Ld5_D.f09_g17.I1850Clm50BgcCropCmip6.cheyenne_intel.clm-basic`:
  shortened to `SMS_Ld2_D`.
  
- `ERS_D_Ld5.f09_g17.I2000Clm45Fates.cheyenne_intel.clm-FatesColdDef`:
  removed (as per discussion in
  https://github.com/ESCOMP/ctsm/issues/275)
  
- `ERS_D_Ld5.f09_g17.I2000Clm50Fates.cheyenne_intel.clm-FatesColdDef`:
  shortened and made non-debug (`ERS_Ld3`) (as per discussion in
  https://github.com/ESCOMP/ctsm/issues/275)

billsacks commented on bld/namelist_files/namelist_defaults_ctsm.xml:
```diff
@@ -474,11 +474,11 @@ attributes from the config_cache.xml file (with keys converted to upper-case).
 -->
 <init_interp_attributes sim_year="1850" use_cndv=".false." use_fates=".false." lnd_tuning_mode="clm4_5_GSWP3v1"
->hgrid=0.9x1.25 maxpft=17 mask=gx1v6 use_cn=.true. use_nitrif_denitrif=.true. use_vertsoilc=.true. use_crop=.false. irrigate=.false. glc_nec=10
+>hgrid=0.9x1.25 maxpft=17 mask=gx1v7 use_cn=.true. use_nitrif_denitrif=.true. use_vertsoilc=.true. use_crop=.false. irrigate=.true. glc_nec=10
```
Here and in the entry below, as well as in the corresponding finidat entries, note that I changed `irrigate=.false.` to `irrigate=.true.`. It appears that this was incorrect before, and that irrigate was `.true.` on the original finidat files pointed to by these configurations. See the log message for commit 7d59fb160 for details.

ekluzek replied:
Yes, this makes sense, since irrigate=T for SP cases. So it looks like it was wrong before. And you changed the mask to gx1v7, since the new files are at gx1v7.

billsacks commented:
@ekluzek please especially see the notes on answer changes in my original PR comment, and let me know if you're comfortable with this.

ekluzek commented on bld/namelist_files/namelist_defaults_ctsm.xml:
```diff
@@ -516,64 +516,64 @@ attributes from the config_cache.xml file (with keys converted to upper-case).
+         ic_tod="0" glc_nec="10" use_crop=".true." irrigate=".false."
          lnd_tuning_mode="clm5_0_GSWP3v1"
->lnd/clm2/initdata_map/clmi.I1850Clm50BgcCrop.1366-01-01.0.9x1.25_gx1v6_simyr1850_c171213.nc
+>lnd/clm2/initdata_map/clmi.I1850Clm50BgcCrop-ciso.1366-01-01.0.9x1.25_gx1v7_simyr1850_c190116.nc
```
yes that makes sense.

ekluzek commented on bld/namelist_files/namelist_defaults_ctsm.xml:
```diff
@@ -516,64 +516,64 @@ attributes from the config_cache.xml file (with keys converted to upper-case).
-<finidat hgrid="0.9x1.25"  maxpft="17"  mask="gx1v6" use_cn=".true." use_cndv=".false." use_fates=".false."
+<finidat hgrid="0.9x1.25"  maxpft="17"  mask="gx1v7" use_cn=".true." use_cndv=".false." use_fates=".false."
          ic_ymd="18500101" use_nitrif_denitrif=".true." use_vertsoilc=".true." sim_year="1850"
-         ic_tod="0" glc_nec="10" use_crop=".false." irrigate=".false." use_init_interp=".true."
```
This looks right to me. Since, the files now don't require interpolation, you leave off use_init_interp=T for each file, and let it only set it if it isn't an exact match. This section here says you can do interpolation if CNDV is off, FATES is off, and the sim year is 1850 or 2000. In other cases you can't do interpolation.

ekluzek commented on doc/checklists/interpolating_initial_conditions.md:
```diff
@@ -0,0 +1,58 @@
+# Steps for interpolating out-of-the-box initial conditions
```
Hmmm. I'm not sure I see the idea of having a directory for checklists? I suppose if they go under docs it makes sense. But, in some cases shouldn't the checklist be IN the directory where it applies? The checklist for making surface datasets would then be in the tools/mksurfdata_map directory, because that's where you have to go to do that task. It's not really natural to think "I should look in the checklists directory for a checklist regarding a task that I don't know there even might be a checklist for". The checklist for bringing changes to master belongs in doc, since that's where you make the ChangeLog.

I think what I would prefer is a naming convention that makes it obvious that something is a checklist that should be followed. For example, something like...

README.CHECKLIST.<description>

So it brings your attention to it, says it's a checklist, and tells you what it's regarding...

ekluzek replied:
Based on the location where this comes into play, I'd say this checklist naturally goes under the bld/namelist_files directory right now.

billsacks replied:
I'm happy with your idea of putting checklists where they most logically belong, but with a naming convention. In the case of this file, I don't think I'd remember that it's in bld/namelist_files, but if we have a convention that all such checklists should have CHECKLIST in their name, then I can at least do `find . -name '*CHECKLIST*'` to find it.

So what would you recommend this file be called? Are you thinking `README.CHECKLIST.interpolating_initial_conditions.md`, for example?

ekluzek replied:
Yes, exactly that's what I'm thinking. I don't know that CHECKLIST has to be in caps, but it makes it yell louder that way, and the standard README. as a prefix also brings attention to it. We could also put a note at the top that such checklists are named that way.

ekluzek commented:
I'm not concerned about the answer changes here, since they are just for initialization, and they are small. Since, we do long simulations small changes at initialization really just aren't going to matter. As you point out we expect changes around the Caspian Sea, but since the domain files are different and updated, there's likely other small changes throughout the globe. And that's likely just because a different machine was used to create the mapping and domain files. So really in the end this is expected. If you compared the domain and mapping files you might see the same pattern as my guess regarding it.

billsacks commented:
@ekluzek I have pushed a change that (I think) addresses your comments about the organization of checklists. Note that I have moved and renamed CTSMMasterChecklist, too, following your suggestion (if I understood you correctly).

ekluzek commented:
Yes, awesome. Looks good.

billsacks commented:
In my latest run of the test suite, with the changes in this PR, the test suite cost was about 4300 core-hours (down from the previous 7700 core-hours). This is getting close to what I'd consider an acceptable cost. (Though note that there might be some machine variability: these numbers are each from a single run of the test suite.)
